Honeywell fixes and improvements

dansarginson opened the pull request:
## Description:
Update Honeywell ClimateDevice to have a meaningful state and to raise state_changed events when the state has changed.

**Related issue (if applicable):** fixes #8688 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

dansarginson committed:
Honeywell fixes and improvements

Give the Honeywell device a state ('On', 'Off', etc) that
can be displayed to user and understood by other components.
Previously this was always 'Unknown'. Update also raises a
state_changed event when a new temperature is polled.

These two together fix an issue (#8688) where Honeywell
climate data couldn't be logged in InfluxDB.

MartinHjelmare commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -75,7 +76,7 @@ def _setup_round(username, password, config, add_devices):
         for i, zone in enumerate(zones):
             add_devices(
-                [RoundThermostat(evo_api, zone['id'], i == 0, away_temp)]
+                [RoundThermostat(hass, evo_api, zone['id'], i == 0, away_temp)]
```
Call `add_devices` with a second argument `True`. This means that `update` will be called directly after adding the entity. Then you should remove the call to `update` from the `__init__` method of `RoundThermostat`.
```py
add_devices([RoundThermostat(evo_api, zone['id'], i == 0, away_temp)], True)
```

dansarginson replied:
Done

MartinHjelmare commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -210,6 +212,24 @@ def update(self):
+
+        device_data = list(
+            filter(lambda x: x['name'] == self.name,
+                   self.device.full_data['devices']))
```
If I'm reading the library code correctly my suggestion below should work and be a lot cleaner. We should consider renaming the instance attribute `self.device` to `self.client` on `RoundThermostat`. Having it be named `self.device` is very confusing when it's just an instance of the client, and there are other things better named devices.
```py
device = self.device.named_devices[self.name]
self.device.system_mode = device['thermostat']['changeableValues']['mode']
```

MartinHjelmare commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -210,6 +212,24 @@ def update(self):
+                device_data[0]['thermostat']
+                ['changeableValues']['mode'])
+
+        self.hass.async_add_job(self.async_update_ha_state())
```
Remove this. When home assistant state is updated after a service call, `update` (this method) on entities will be called automatically for polling entities.

dansarginson replied:
Oh, apologies, I'd misunderstood this completely.

MartinHjelmare commented:
We should set a list of supported operation modes in the `operation_list` property for the entity too.

dansarginson committed:
Roll back some changes

These were not necessary to achieve the result I wanted.

dansarginson commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -115,8 +116,9 @@ def _setup_us(username, password, config, add_devices):
     """Representation of a Honeywell Round Connected thermostat."""
 
-    def __init__(self, device, zone_id, master, away_temp):
+    def __init__(self, hass, device, zone_id, master, away_temp):
```
Ok, understood

dansarginson commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -210,6 +212,24 @@ def update(self):
             self._name = data['name']
             self._is_dhw = False
 
+        if self.entity_id is None:
```
I was, but now I see that it was my error. Apologies. Removed.

dansarginson committed:
Renamed RoundThermostat's 'device' member for greater clarity

Now called 'client'

dansarginson committed:
Improve and simplify discovering thermostat mode

Per code review, this is a rather neater way to discover the thermostat mode

dansarginson committed:
Update tests for compatibility with new component

The tests previously relied upon the update() method being
called in the constructor. This is no longer the case.

dansarginson commented:
@MartinHjelmare I think I've made all the changes you wanted. Please could you take a look? It's much cleaner and simpler now, thanks.

MartinHjelmare commented on homeassistant/components/climate/honeywell.py:
```diff
@@ -210,6 +210,13 @@ def update(self):
 
+        # The underlying library doesn't expose the thermostat's mode
+        # but we can pull it out of the big dictionary of information.
+        device = self.client.named_devices[self.name]
```
I'm thinking about the case when the device is a "DOMESTIC_HOT_WATER" thermostat. Since we set `self._name` to 'Hot Water' in that case, I don't think we can use the name to get the device in that case. We could use the id instead, which will work for all cases.
```py
device = self.client.devices[self._id]
```
But will a "DOMESTIC_HOT_WATER" device have `device['thermostat']['changeableValues']['mode']` accessible?

dansarginson replied:
I don't think we can assume that 'mode' will be available. Unfortunately I can't test as I don't have a DHW device.

MartinHjelmare replied:
Good suggestion! :+1:

dansarginson replied:
Done!

MartinHjelmare reviewed (changes requested):
I forgot about the case when the device is a "DOMESTIC_HOT_WATER" thermostat. See question below. Also a small formatting comment.

dansarginson committed:
Address formatting review comment

Parens not necessary

dansarginson committed:
This system mode is not certain to apply to domestic hot water

Moved the mode lookup to only happen on update of radiator devices,
since hot water devices seem to be treated differently and I can't test.

MartinHjelmare reviewed (approved):
Nice! :tada: 

> We should set a list of supported operation modes in the operation_list property for the entity too, and update set_operation_mode to use the supported client methods to change mode.

This can be adressed in a future PR.
